### PR TITLE
Median imputation and fillNA with zero

### PR DESCRIPTION
**Pull request by eftekhari-mhs**
FillMean is changed to FillNA(TimeSeries, method)
Implemented methods: mean, zero, median

### CODE REVIEW

File: imputation/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Go module for adding missing dates of a time series with value = mean
```
Comment (jeffreyoldham) [inline imputation/README.md:1]:
```suggestion
Go module for adding missing values to a time series.
```

File: imputation/ts_impute.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	yearfmt  = "2006"
+)
+
+// The diff function returns the minimum gap (will change to GCD) from a given *sorted* set of data points
+// that all have the same format.
+// TODO(eftekhari-mhs): add OK, ERR return values and handle errors.
+func diff(keys []string, dateFormat string) (int, int, int) {
+	year, month, day := 0, 0, 0
+
+	duration := 1000 //A large number.
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:36]:
```suggestion
	duration := 1000 // A large number.
```

File: imputation/ts_impute.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	case "mean":
+		for _, v := range values {
+			fillV += v
+		}
+		fillV = fillV / float64(len(keys))
+	case "zero":
+		fillV = 0
+	case "median":
+		sort.Float64s(values)
+		fillV = values[len(values)/2]
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:84]: Median is defined as the average of the two middle elements if there are an even number of elements.  Q.v., https://en.wikipedia.org/wiki/Median#Formal_definition.

File: imputation/ts_impute.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		parseFormat = yearfmt
+	case 7:
+		parseFormat = monthfmt
+	case 10:
+		parseFormat = dayfmt
+	}
+
+	if parseFormat == "" {
+		return ts, errors.New("date format is not ISO 8601")
+	}
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:101]:
```suggestion
        default:
          return ts, errors.New("date format is not ISO 8601")
	}
```

This suggestion is purely a matter of taste.  It is pretty arbitrary.

File: imputation/ts_impute_test.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+// computeMedian computes the median value of the given non-empty time series's data.
+func computeMedian(s TimeSeries) float64 {
+	values := make([]float64, 0)
+	for _, v := range s {
+		values = append(values, v)
+	}
+	sort.Float64s(values)
+	return values[len(values)/2]
```
Comment (jeffreyoldham) [inline imputation/ts_impute_test.go:25]: Fix this also please.

File: imputation/ts_impute.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+			year = int(y2 - y1)
+			month = int(M2 - M1)
+			day = int(d2 - d1)
+		}
+	}
+
+	return year, month, day
+}
+
+// FillNA returns TimeSeries with missing datapoints with assigned values according to the selected method.
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:57]: `FillNA` is probably meant to mean "fill the NA values".  However, I read it as "fill with NA values".  I wonder what the best function name is.

File: imputation/ts_impute_test.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+// createOutput generate a new time series by adding desiredValues as data points and replace
+// their value by mean.
```
Comment (jeffreyoldham) [inline imputation/ts_impute_test.go:29]:
```suggestion
// their value by using the specified algorithm.
```

**Review by jeffreyoldham — APPROVED** [review]
Thanks.  After you make the suggested changes, please ask Tiffany and perhaps me for a review so we can merge.

File: imputation/ts_impute.go (hunk 6 above)
Comment (eftekhari-mhs) [reply imputation/ts_impute.go:57]: I looked what Pandas use and I used the same name! :D

File: imputation/ts_impute.go [hunk 8]
```diff
@@ -81,7 +81,14 @@ func FillNA(ts TimeSeries, method string) (TimeSeries, error) {
 		fillV = 0
 	case "median":
 		sort.Float64s(values)
-		fillV = values[len(values)/2]
+        if len(values) % 2 == 0{
+            // In case of even length: median = average of two middle values.
+            fillV = (values[len(values)/2] + values[len(values)/2 -1])/2
+            log.Printf("Median is %v: for even case" , fillV)
+        }else{
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:88]:
```suggestion
        } else {
```

**Review by jeffreyoldham — APPROVED** [review]
Thanks.

File: imputation/ts_impute_test.go [hunk 9]
```diff
@@ -55,12 +116,19 @@ func createOutput(ts TimeSeries, desiredValues []string, method string) TimeSeri
...
+				out[k] = series[k]
+			}
+		}
 	}
 	return (out)
 }
 
 func TestFillNA(t *testing.T) {
-	methods := []string{"mean", "zero", "median"}
+	methods := []string{"mean", "zero", "median", "spline1"}
```
Comment (jeffreyoldham) [inline imputation/ts_impute_test.go:131]: "spline1" is implemented by Interpolate, not FillNA, so it should be tested in a function with a name like `TestInterpolate`.

File: imputation/ts_impute_test.go [hunk 10]
```diff
@@ -22,11 +23,71 @@ func computeMedian(s TimeSeries) float64 {
...
-    }
+	if len(values)%2 == 0 {
+		return (values[len(values)/2] + values[len(values)/2-1]) / 2
+	} else {
+		return values[len(values)/2]
+	}
+}
+
+// Values are produced using Python, Pandas linear interpolation method:
+func linearInterpolate(date string) TimeSeries {
```
Comment (jeffreyoldham) [inline imputation/ts_impute_test.go:34]: Is this a good function name?

File: imputation/ts_impute_test.go [hunk 11]
```diff
@@ -22,11 +23,71 @@ func computeMedian(s TimeSeries) float64 {
...
+
+	// "length 3, years"
+	yseries := TimeSeries{
```
Comment (jeffreyoldham) [inline imputation/ts_impute_test.go:37]: We should test the "zero" and "one" cases, too.

**Review by jeffreyoldham — COMMENTED** [review]
Thanks.

File: imputation/ts_impute.go [hunk 12]
```diff
@@ -51,7 +61,23 @@ func diff(keys []string, dateFormat string) (int, int, int) {
...
+	return keys
+}
+
+func getValues(ts TimeSeries) []float64 {
+	values := make([]float64, 0)
+	for _, v := range ts {
+		values = append(values, v)
+	}
+	return values
 }
```
[the remark this answers is not in the document]
Comment (eftekhari-mhs) [reply imputation/ts_impute.go:81]: For interpolation I'm not using values, this is why I separated them.

File: imputation/ts_impute.go [hunk 13]
```diff
@@ -51,7 +61,23 @@ func diff(keys []string, dateFormat string) (int, int, int) {
 		}
 	}
 
-	return year, month, day
+	return parseFormat, year, month, day
+}
+
+func getKeys(ts TimeSeries) []string {
+	keys := make([]string, 0)
```
[the remark this answers is not in the document]
Comment (eftekhari-mhs) [reply imputation/ts_impute.go:68]: If I set a length (e.g. to 5) then appending will add to the end of the array and assume the first five values are "". 

keys := make([]string, 5)
    for i := 0; i < 5; i++ {
        keys = append(keys, "a")
    }

result: [     a a a a a a]

File: imputation/ts_impute_test.go (hunk 11 above)
Comment (eftekhari-mhs) [reply imputation/ts_impute_test.go:37]: They are being tested, just that these cases don't need precomputed outputs.

File: imputation/ts_impute_test.go (hunk 9 above)
Comment (eftekhari-mhs) [reply imputation/ts_impute_test.go:131]: Right! I changed the testing structure. Now it'll be easy to add test values for quadratic and cubic ones. There is a structure (InterpolateValue) that I can easily add more to its values.

File: imputation/ts_impute.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jeffreyoldham) [reply imputation/ts_impute.go:75]: I apologize for my mistake.  Use `make([]string, 0, 5)`.  I read https://golang.org/pkg/builtin/#make.

File: imputation/ts_impute_test.go [hunk 14]
```diff
@@ -116,19 +61,31 @@ func createOutput(ts TimeSeries, desiredValues []string, method string) TimeSeri
 				out[k] = median
 			}
 		}
-	case "spline1":
+	}
+	return (out)
+}
+
+// createOutputInterpolate generate a new time series by adding desiredValues as data points and replace
+// their value using given precomputed values for each method.
```
Comment (jeffreyoldham) [inline imputation/ts_impute_test.go:69]:
```suggestion
// createOutputInterpolate generates a new time series by adding desiredValues as data points and replaces
// their value using given precomputed values for each method.
```

**Review by jeffreyoldham — APPROVED** [review]
Thanks.

Let's get this merged.

File: imputation/ts_impute.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (tjann) [reply imputation/ts_impute.go:75]: @eftekhari-mhs do you plan to make this change?

File: imputation/ts_impute.go (hunk 6 above)
Comment (tjann) [reply imputation/ts_impute.go:91]: Fine with me.

File: imputation/ts_impute.go [hunk 15]
```diff
@@ -20,28 +20,44 @@ import (
...
+// Args:
+//   keys: A sorted list of times all with the same format, one of those listed above.
+// Returns:
+//   1) The time format of the keys.
+//   2) The minimum gap between the keys in terms of years, months, and days.
+// TODO(eftekhari-mbs): Modify the algorithm to compute a greatest common divisor-like algorithm, not
+// just the minimum gap.
 // TODO(eftekhari-mhs): add OK, ERR return values and handle errors.
-func diff(keys []string, dateFormat string) (int, int, int) {
+func diff(keys []string) (string, int, int, int) {
```
Comment (tjann) [inline imputation/ts_impute.go:40]: Could consider renaming this function to be more specific

File: imputation/ts_impute.go [hunk 16]
```diff
@@ -52,51 +68,134 @@ func diff(keys []string, dateFormat string) (int, int, int) {
...
+		if len(values)%2 == 0 {
+			// In case of even length: median = average of two middle values.
+			fillV = (values[len(values)/2] + values[len(values)/2-1]) / 2
+			log.Printf("Median is %v: for even case", fillV)
+		} else {
+			fillV = values[len(values)/2]
+			log.Printf("Median is %v: for odd case", fillV)
+		}
+	default:
+		return ts, errors.New("unknown method")
```
Comment (tjann) [inline imputation/ts_impute.go:130]: Can you make the error msgs consistent in formatting? I'm seeing mostly but not all complete(ish) sentences

File: imputation/ts_impute.go [hunk 17]
```diff
@@ -52,51 +68,134 @@ func diff(keys []string, dateFormat string) (int, int, int) {
...
+			ts[fmt.Sprint(d.Format(parseFormat))] = fillV
+		}
+	}
+	return ts, nil
+}
+
+// Interpolate function's input is a TimeSeries with possible missing datapoints.
+// The function returns a Timeseries with no missing dates. Interpolation method is used to fill the values.
+// Methods can vary by degree: d = {1} is implemented.
+func Interpolate(ts TimeSeries, degree int) (TimeSeries, error) {
```
Comment (tjann) [inline imputation/ts_impute.go:148]: It's not clear to me that FillNA and Interpolate should be separate functions. Could you convince me?

File: imputation/ts_impute_test.go [hunk 18]
```diff
@@ -14,102 +21,260 @@ func computeMean(s TimeSeries) float64 {
...
+				out[k] = median
+			}
+		}
+	}
+	return (out)
+}
+
+// createOutputInterpolate generates a new time series by adding desiredValues as data points and replaces
+// their value using given precomputed values for each method.
+func createOutputInterpolate(ts TimeSeries, desiredValues map[string]InterpolateValues, method string) TimeSeries {
```
Comment (tjann) [inline imputation/ts_impute_test.go:70]: I wonder if this should be called createExpectedTimeSeries or similar?

File: imputation/ts_impute_test.go [hunk 19]
```diff
@@ -14,102 +21,260 @@ func computeMean(s TimeSeries) float64 {
...
-				"2002-02-01": 18.9,
-				"2002-02-02": 28,
+			"length 3, years": {
+				series: TimeSeries{
+					"2003": 42.5,
+					"2002": 18.9,
+					"1998": 28,
+					"1996": 25,
+				},
+				additionalKeys: []string{"1999", "2000", "2001", "1997"},
```
Comment (tjann) [inline imputation/ts_impute_test.go:109]: Very nice.

File: imputation/ts_impute_test.go [hunk 20]
```diff
@@ -14,102 +21,260 @@ func computeMean(s TimeSeries) float64 {
...
+					"2002-02-23": {linear: 39.277},
+					"2002-02-24": {linear: 39.814},
+					"2002-02-25": {linear: 40.351},
+					"2002-02-26": {linear: 40.888},
+					"2002-02-27": {linear: 41.425},
+					"2002-02-28": {linear: 41.962},
+				},
+			},
+		}
+		for label, test := range tests {
```
Comment (tjann) [inline imputation/ts_impute_test.go:260]: Nice parameterization of tests.

File: imputation/ts_impute.go (hunk 17 above)
Comment (eftekhari-mhs) [reply imputation/ts_impute.go:148]: FillNA has two inputs: time series, method = "mean", "median" but not a degree (integer)
for Interpolate we have one more input besides the method name, for now method name is assumed to be "spline" but some might consider adding "polynomial" or some other methods.
Maybe you're thinking of merging these two methods together and somehow managing one interface with two implementation? one that accept degree (int) and the other not.

Comment (tjann) [reply imputation/ts_impute.go:148]: Or degree can be an optional argument. This could be an optimization for later, you can ask Jeffrey what he thinks.

**Review by tjann — APPROVED** [review]

File: imputation/ts_impute.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (tjann) [reply imputation/ts_impute.go]: When you call this function for interpolation, could you just discard the values?

`keys, _ = ...`

**Comment by tjann** [comment]
Will merge the PR, but please TAL at the unresolved comments for next round.

File: imputation/ts_impute.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (eftekhari-mhs) [reply imputation/ts_impute.go:75]: yes, I'll do it in the next round.

**Pull request merged**